contracts/governance/Vesting/fouryear/FourYearVestingLogic.sol

```diff
@@ -0,0 +1,296 @@
+	}
+
+	/**
+	 * @notice Allows the owners to migrate the positions
+	 * to a new staking contract.
+	 * */
+	function migrateToNewStakingContract() external onlyOwners {
+		staking.migrateToNewStakingContract();
+		staking = Staking(staking.newStakingContract());
+		emit MigratedToNewStakingContract(msg.sender, address(staking));
+	}
+
+	/**
+	 * @notice Extends first year stakes for four year vesting contracts.
+	 * @dev Tokens are vested for 4 years. Since the max staking
+	 * period is 3 years and the tokens are unlocked only after the first year is
+	 * passed, hence, we extend the duration of staking for all unlocked tokens for the first
+	 * year by 3 years.
+	 * */
+	function extendStaking() external {
+		uint256 oneYear = startDate.add(52 weeks);
+		uint256[] memory dates;
+		uint96[] memory stakes;
+		(dates, stakes) = staking.getStakes(address(this));
+
+		for (uint256 i = 0; i < dates.length; i++) {
+			if ((dates[i] < block.timestamp) && (dates[i] <= oneYear) && (stakes[i] > 0)) {
+				staking.extendStakingDuration(dates[i], dates[i].add(156 weeks));
+			}
+		}
+	}
+
+	/**
+	 * @notice Stakes tokens according to the vesting schedule. Low level function.
+	 * @dev Once here the allowance of tokens is taken for granted.
+	 * @param _sender The sender of tokens to stake.
+	 * @param _amount The amount of tokens to stake.
+	 * @param _restartStakeSchedule The time from which staking schedule restarts.
+	 * The issue is that we can only stake tokens for a max duration. Thus, we need to restart
+	 * from the lastSchedule.
+	 * @return lastSchedule The max duration for which tokens were staked.
+	 * @return remainingAmount The amount outstanding - to be staked.
+	 * */
+	function _stakeTokens(
+		address _sender,
+		uint256 _amount,
+		uint256 _restartStakeSchedule
+	) internal returns (uint256 lastSchedule, uint256 remainingAmount) {
+		// Creating a new staking schedule for the same vesting contract is disallowed unlike normal vesting
+		require((startDate == 0) || (startDate > 0 && remainingStakeAmount > 0 && _restartStakeSchedule > 0), "create new vesting address");
+		uint256 restartDate;
+		uint256 relativeAmount;
+		// Calling the _stakeTokens function first time for the vesting contract
+		// Runs for maxInterval only (consider maxInterval = 18 * 4 = 72 weeks)
+		if (startDate == 0 && _restartStakeSchedule == 0) {
+			startDate = staking.timestampToLockDate(block.timestamp); // Set only once
+			durationLeft = duration; // We do not touch duration and cliff as they are used throughout
+			cliffAdded = cliff; // Hence, durationLeft and cliffAdded is created
+		}
+		// Calling the _stakeTokens second/third time - we start from the end of previous interval
+		// and the remaining amount(amount left after tokens are staked in the previous interval)
+		if (_restartStakeSchedule > 0 && _restartStakeSchedule == lastStakingSchedule && _amount == remainingStakeAmount) {
```

Comment:
For incorrect _restartStakeSchedule or _amount, the restart date will be set to start date which is incorrect.